### ISSUE TITLE
Why Gumbel-Softmax relaxation applied in sample generation?

### ISSUE THREAD

**Issue by ruoyuli1995** [weilinie/RelGAN#17]
Hi, thank you for sharing the code!
But I have a question for sample generation... In my opinion, Gumbel-Softmax relaxation is used for directly passing gradients back to generator. However, why it is still applied when generating samples?
Thanks ahead!

**Comment by weilinie** [comment]
Because we replace the sampling with the Gumbel trick. We keep using the Gumbel trick instead of sampling directly during sample generation just in order for maintaining the same pipeline between training and testing. Otherwise, we may loss randomness.

**Comment by ruoyuli1995** [comment]
Thank you for your reply!
I am still a little confused about "randomness".
Now I trained a model on coco data with a high temperature, but if I use it for generation, should I still keep a high temperature? It seems to generate some bigrams that are illogical.
Maybe I should express in a another way, I can understand the sample process from softmax which still matches the inner distribution of texts, but I cannot understand the randomness from Gumbel trick, especially when I use a high temperature.

**Comment by weilinie** [comment]
Sorry for having not made it clear. Be aware that Gumbel trick and softmax approximation are two separate parts in Gumbel-Softmax. When I said “we keep using the Gumbel trick instead of sampling directly during sample generation”, I didn’t mention we still use the softmax part for sample generation. You can see the [code](https://github.com/weilinie/RelGAN/blob/993faa2a82dee181a0aeedfaa43fd66f4d4fb198/models/lstm_vanilla.py#L31), where `next_token` for the sampling.

**Comment by ruoyuli1995** [comment]
Okay, I understand now. Thank you so much for your time and your patient explanation!
I will close this issue then. :)
